src/osd/PrimaryLogPG.cc

```diff
@@ -8372,15 +8340,87 @@ void PrimaryLogPG::write_update_size_and_usage(object_stat_sum_t& delta_stats, o
     oi.size = new_size;
   }
  
```

Comment:
This portion is too complicated to leave here.  Try to find a way to factor out the complex portion into a helper in osd_types.h with appropriate tests of just that helper.  Be sure to include a precise description of what the helper does.

Reply:
Also, I don't think that dropping references in write_update_size_and_usage is a good idea -- readers seeing that method invocation wouldn't expect rados operations to result.  I do like the idea of doing it once at the end based on the modified map, however -- just do it in a seperate method which invokes the helper I mentioned above for the complex logic.